test/metabase/models/login_history_test.clj

```diff
@@ -47,7 +48,12 @@
                                                            (into {} (for [ip-address ip-addresses]
                                                                       [ip-address
                                                                        {:description "San Francisco, California, United States"
-                                                                        :timezone    (t/zone-id "America/Los_Angeles")}])))]
+                                                                        :timezone    (t/zone-id "America/Los_Angeles")}])))
+                          login-history/maybe-send-login-from-new-device-email
+                          (fn [login-history]
+                            (when-let [futur (original-maybe-send login-history)]
+                              ;; block in tests
+                              (deref futur)))]
```

Comment:
If this never gets called (e.g. because of an Exception that gets thrown before the future is called) the tests are going to hang forever... and it's going to be almost impossible to debug the reason why. You should use `u/deref-with-timeout` here instead. 30 seconds or so is probably fine... normally it should be fast but you never know with those slow CircleCI test runners and their GC pauses

Reply:
thanks. really good catch